gpytorch/utils/sparse.py

```diff
@@ -84,10 +83,31 @@ def bdsmm(sparse, dense):
     """
     Batch dense-sparse matrix multiply
     """
+    # Make the batch sparse matrix into a block-diagonal matrix
     if sparse.ndimension() > 2:
-        batch_size, num_rows, num_cols = sparse.size()
-        batch_assignment = sparse._indices()[0]
-        indices = sparse._indices()[1:].clone()
+        # Expand the tensors to account for broadcasting
+        output_shape = _matmul_broadcast_shape(sparse.shape, dense.shape)
+        expanded_sparse_shape = torch.Size(output_shape[:-2] + sparse.shape[-2:])
+        unsqueezed_sparse_shape = [1 for _ in range(len(output_shape) - sparse.dim())] + list(sparse.shape)
+        repeat_sizes = tuple(
+            output_size // sparse_size
+            for output_size, sparse_size in zip(expanded_sparse_shape, unsqueezed_sparse_shape)
+        )
+        sparse = sparse_repeat(sparse, *repeat_sizes)
+        dense = dense.expand(*output_shape[:-2], dense.size(-2), dense.size(-1))
+
+        # Figure out how much need to be added to the row/column indices to create
+        # a block-diagonal matrix
+        *batch_shape, num_rows, num_cols = sparse.shape
+        batch_size = torch.Size(batch_shape).numel()
+        batch_multiplication_factor = torch.tensor(
+            [torch.Size(batch_shape[i + 1:]).numel() for i in range(len(batch_shape))],
+            dtype=torch.long, device=sparse.device
+        )
+        batch_assignment = sparse._indices()[:-2].t() @ batch_multiplication_factor
```

Comment:
on the GPU, this errors out with the following: 
```
[...]

/data/users/balandat/fbsource/fbcode/buck-out/dev/gen/bento/kernels/bento_kernel_ae_lazarus#link-tree/gpytorch/utils/sparse.py in bdsmm(sparse, dense)
    105             dtype=torch.long, device=sparse.device
    106         )
--> 107         batch_assignment = sparse._indices()[:-2].t() @ batch_multiplication_factor
    108 
    109         # Create block-diagonal sparse tensor

RuntimeError: addmv for CUDA tensors only supports floating-point types. Try converting the tensors with .float() at caffe2/aten/src/THC/generic/THCTensorMathBlas.cu:150
```

You can either do all of this index business on the cpu, or just add a little if statement in to perform the computation as float on the GPU and then cast back to long (I'm sure there is a more elegant way of doing this...):
```
if sparse.is_cuda:
    batch_assignment = (sparse._indices()[:-2].t().float() @ batch_multiplication_factor.float()).to(dtype=torch.long)
else:
    batch_assignment = sparse._indices()[:-2].t() @ batch_multiplication_factor
```

Reply:
Whoops sorry my comment was stupid. Doing it your way 😁